Add error handling.

tpetersen0308 opened the pull request:
- Respond with 500 status when internal processes throw an IOException.
- Respond with 400 status when request causes an exception to be thrown.
- Add logging for checked exceptions that cannot be handled at runtime.
- Validate port and directory input when starting the server with executable.
- Also registered app.Main in POM file for standard java -jar execution.
- Also renamed files to decrease redundancy and increase specificity.
- Also updated readme.

tpetersen0308 committed:
Add error handling.
- Respond with 500 status when internal processes throw an IOException.
- Respond with 400 status when request causes an exception to be thrown.
- Add logging for checked exceptions that cannot be handled at runtime.
- Validate port and directory input when starting the server with executable.
- Also registered app.Main in POM file for standard java -jar execution.
- Also renamed files to decrease redundancy and increase specificity.
- Also updated readme.

ssunday commented on src/main/java/server/Client.java:
```diff
@@ -1,36 +1,76 @@
 
-    public BufferedReader getInputStreamReader() {
-        return inputStreamReader;
+    public String read() throws IOException{
```
bracket spacing

zacholauson commented on src/main/java/server/ConnectionHandler.java:
```diff
@@ -0,0 +1,45 @@
+        } catch (IOException err) {
+            response = selector.selectResponse(err);
+        }
+        catch (NullPointerException err) {
```
Where does this `NullPointerException` come from? We should be able to handle it without causing an exception if we can anticipate it. https://web.archive.org/web/20140430044213/http://c2.com/cgi-bin/wiki?DontUseExceptionsForFlowControl

tpetersen0308 replied:
`parse()` throws a `NullPointerException` if it receives an improperly formatted request (i.e. Chrome sends an empty request, so `parseRequestMethod()` tries to call `split()` on a null `requestLine`).

tpetersen0308 replied:
But I suppose that pattern is what that post says is a violation of the principle of least astonishment?

tpetersen0308 replied:
Would it be more appropriate to do something like this, since `parseRequestMethod()` is where the exception originates?
```    private String parseRequestMethod(String requestLine) {
        try {
            return requestLine.split(" ")[0].trim();
        } catch (NullPointerException err) {
            return "";
        }
    }```

zacholauson replied:
Yup, I wouldn't expect to have to handle a NullPointerException when parsing a Request. If something can be null, we should check it and handle it appropriately. I could see maybe an IllegalArgumentException if you were provided null, but in this case the null is coming from how the request parsing has been implemented.

tpetersen0308 replied:
Cool, yeah I just refactored so that `parseRequestLine()` won't return `null`, but an empty request will still cause a 400 status.

ssunday commented on src/main/java/server/response/HeaderHelpers.java:
```diff
@@ -0,0 +1,44 @@
+    private static String fileExtension(File file) {
+        String name = file.getName();
+        int lastIndexOf = name.lastIndexOf(".");
+        if (lastIndexOf == -1) {
```
why brackets here and not above for the if statements?

tpetersen0308 replied:
No good reason at all! I've removed them, thanks.

tpetersen0308 committed:
Refactor
- Render a file given a sub-path of default directory or absolute path.
- Handle null pointers in RequestParser instead of throwing exception.
- Also add AlternateRoutes to demonstrate decoupling of server/application code.

zacholauson commented on src/main/java/server/request/Parser.java:
```diff
@@ -23,23 +26,20 @@ public Request parse() {
-        return client.read();
+    private String parseRequestLine() throws IOException {
+        String requestLine = client.read();
+        return requestLine == null ? "" : requestLine;
```
In general I prefer to be explicit about failures like this, rather than return a value that allows an odd state to exist. In other words, from the interface level it's not overly clear that this may return an empty string, or what that even means in this context. Eventually that empty string makes it's way into the Selector after some additional parsing, but that's pretty far from the source. 

One option would be to use `java.util.Optional`s to be explicit about the lack of real value here. While it would be similar to returning null, it allows the caller to optionally act on the value if it is present, otherwise it becomes a NOOP. I started playing around with this, but it doesn't quite fit into the current handler/selector structure. I'm fine with this for now, just wanted to call that out. 

Outside of this specific feedback I'd recommend looking into Optionals if you haven't already, they can be very powerful, and in my experience the readability of the code is quite nice compared to null/empty checks all over.

tpetersen0308 replied:
To me it made sense to just assign an empty string to a `Request` field in the case that the request is empty and use that value to determine what kind of response to send--do you think the solution here could be as simple as introducing a meaningfully named temp variable and return something like `emptyRequestLine` to clarify what's happening?

zacholauson replied:
That's fair. I guess I was wondering about reducing the scope in which that empty string context exists. With your explanation I think it makes sense in the context of the Parser, but maybe less so in the Selector. I'm fine with this for now.

zacholauson commented on src/main/java/server/ConnectionHandler.java:
```diff
@@ -1,27 +1,35 @@
+        try {
+            request = parser.parse();
+            response = selector.selectResponse(request);
+        } catch (IOException err) {
```
While it's nice that this results in an internal server error, it's probably worth still logging this exception somewhere for debugging purposes. In a more realistic example, you'd definitely want to know what's going on internally if your service started spitting out tons of internal server errors.

zacholauson commented on src/main/java/server/Server.java:
```diff
@@ -11,25 +13,26 @@
-                RequestParser parser = new RequestParser(client);
-                ResponseSelector selector = new ResponseSelector(routes);
-                ResponseFormatter formatter = new ResponseFormatter();
+                Parser parser = new Parser(client);
```
This is a fairly small thing, but ConnectionHandler has both `client` and a `parser` with the same `client`. Do you think it would make sense to remove the `client` state from the `parser` and instead have the `ConnectionHandler` pass the client to `parser#parse`?

As a side note, if you removed the request specific state from the `parser`, I think you could reuse the `parser`, `selector,` and `formatter` instead of instantiating a new one for every request.

tpetersen0308 replied:
Yes, I think that would make a lot of sense! Do you think the temporal dependency between `parseRequestLine`, `parseRequestHeaders`, and `parseRequestBody` could create any vulnerability to concurrency bugs?

zacholauson replied:
Good question, that's important to think about. Based on the current structure of `Parser`, what do you think?

tpetersen0308 replied:
I want to say it's fine, because the dependent methods are private, and the temporal dependencies are honored by the only public method that calls them (`parse`).

zacholauson replied:
Agreed

zacholauson commented on src/main/java/server/directory/DirectoryIndex.java:
```diff
@@ -35,7 +35,7 @@ private String currentFolder() {
 
-        for(File file : sortedContents()) {
+        for (File file : sortedContents()) {
             String href = file.getPath().replace(DefaultDirectory.PATH, "");
```
I must have missed this in a previous review, but do you think it's possible to remove the global mutable state from `DefaultDirectory` (`PATH`)?

zacholauson commented:
I think the abstractions are looking much better now, nice work @tpetersen0308.

tpetersen0308 committed:
Refactor
- Add logging to all catch blocks.
- Remove client field from Parser and pass in as method parameter,
  to allow for using same instances of parser, selector, and formatter
  for each request.
- Make default path immutable once set.

zacholauson commented on src/main/java/server/Server.java:
```diff
@@ -20,14 +20,14 @@ public Server(Integer port) {
     public void start(Map<String, Map<String, Handler>> routes) {
+        Parser parser = new Parser();
+        Selector selector = new Selector(routes);
+        Formatter formatter = new Formatter();
```
Nice!
